File: packages/engine/src/core/rule-fidelity.ts
/**
 * Rule fidelity (memory-lead 1.2, D2) — what a standing rule states that the
 * operator's own words do not.
 *
 * A rule is the agent's restatement of something the operator said, and a
 * restatement can add law nobody made: round 1's claude-sofar wrote "reject
 * anything else with 4xx" for an operator's "Reject anything else", the digest
 * told every later session to obey the rule verbatim, and S9 returned a 400 the
 * tests did not want. The decision now carries the operator's `quote`, and
 * this module names the SPECIFICS the rule adds — the concrete, checkable
 * tokens a paraphrase slips in:
 *
 *  - status: an HTTP status code or class — `400`, `4xx`;
 *  - path:   a token with `/` between path characters (`apps/web`, `/api`),
 *            or a file name with an extension (`categories.ts`);
 *  - value:  a backticked or double-quoted span, or any other token carrying
 *            a digit (`30s`, `6`, `v2`) — decision and memory handles (`D3`,
 *            `M2`) excepted, since those cite the record, not the world.
 *
 * Pure and deterministic: no env, no clock, no locale — the digest renders
 * from it, and rust-core mirrors the digest byte for byte (memory-lead 1.4).
 * Matching is case-insensitive over whitespace-collapsed text, and a term
 * whose edge is alphanumeric must meet a non-alphanumeric boundary in the
 * quote, so `400` is not found inside `4000`.
 */

export type RuleSpecificKind = 'status' | 'path' | 'value'

export interface RuleSpecific {
  kind: RuleSpecificKind
  text: string
}

/** Backticked and double-quoted spans (straight or curly), inner text captured. */
const SPAN_RE = /`([^`]+)`|"([^"]+)"|“([^”]+)”/g
const LEADING_PUNCT = /^[([{<'"‘“]+/
const TRAILING_PUNCT = /[)\]}>'"’”,;:.!?]+$/
const STATUS_RE = /^(?:[1-5][0-9]{2}|[1-5]xx)$/i
const SLASH_PATH_RE = /[\w.~-]\/[\w.*-]|^\/[\w.-]/
const FILE_NAME_RE = /^[\w-]{2,}(?:\.[\w-]+)*\.[A-Za-z][A-Za-z0-9]{0,4}$/
const HANDLE_RE = /^[DM][1-9][0-9]*$/

const collapse = (text: string): string => text.replace(/\s+/g, ' ').trim()

function classify(token: string): RuleSpecificKind | null {
  if (STATUS_RE.test(token)) return 'status'
  if (SLASH_PATH_RE.test(token) || FILE_NAME_RE.test(token)) return 'path'
  if (/[0-9]/.test(token) && !HANDLE_RE.test(token)) return 'value'
  return null
}

/** Every specific the rule states, in rule order, deduplicated case-insensitively. */
export function ruleSpecifics(rule: string): RuleSpecific[] {
  const found: RuleSpecific[] = []
  const seen = new Set<string>()
  const add = (kind: RuleSpecificKind, text: string): void => {
    const key = text.toLowerCase()
    if (text.length === 0 || seen.has(key)) return
    seen.add(key)
    found.push({ kind, text })
  }

  // Spans first, then the text between them, so a span's words are never
  // re-read as tokens and the output stays in the order the rule reads.
  let last = 0
  const tokensOf = (chunk: string): void => {
    for (const raw of chunk.split(/\s+/)) {
      const token = raw.replace(LEADING_PUNCT, '').replace(TRAILING_PUNCT, '')
      const kind = token.length > 0 ? classify(token) : null
      if (kind !== null) add(kind, token)
    }
  }
  for (const m of rule.matchAll(SPAN_RE)) {
    tokensOf(rule.slice(last, m.index))
    add('value', collapse(m[1] ?? m[2] ?? m[3] ?? ''))
    last = m.index! + m[0].length
  }
  tokensOf(rule.slice(last))
  return found
}

const isAlnum = (ch: string | undefined): boolean => ch !== undefined && /[A-Za-z0-9]/.test(ch)

/** `needle` occurs in `haystack` (both lowercased) at a term boundary. */
function containsTerm(haystack: string, needle: string): boolean {
  let at = haystack.indexOf(needle)
  while (at !== -1) {
    const before = isAlnum(needle[0]) && isAlnum(haystack[at - 1])
    const after = isAlnum(needle[needle.length - 1]) && isAlnum(haystack[at + needle.length])
    if (!before && !after) return true
    at = haystack.indexOf(needle, at + 1)
  }
  return false
}

/** The specifics the rule states and the quote does not, as the rule spells them. */
export function unquotedSpecifics(rule: string, quote: string): string[] {
  const words = collapse(quote).toLowerCase()
  return ruleSpecifics(rule)
    .filter((s) => !containsTerm(words, collapse(s.text).toLowerCase()))
    .map((s) => s.text)
}

/**
 * The rule as every surface renders it: verbatim, then — when the decision
 * carries one — the operator's words and what the rule adds to them.
 * Whitespace is collapsed (normalization, not clipping: nothing is cut).
 */
export function renderRule(rule: string, quote: string | undefined): string {
  const text = collapse(rule)
  return quote === undefined ? text : `${text} — ${quoteClause(rule, quote)}`
}

/** `operator: "<quote>"`, plus what the rule adds to it — the half after the rule. */
export function quoteClause(rule: string, quote: string): string {
  const added = unquotedSpecifics(rule, quote)
  const flag = added.length > 0 ? ` (not in the operator's words: ${added.join(', ')})` : ''
  return `operator: "${collapse(quote)}"${flag}`
}

/**
 * The write-time warning (D2): the capturing session learns what its rule
 * adds while the operator's message is still in its context — the one moment
 * a correction costs nothing but the next call. Null when nothing is added.
 */
export function ruleFidelityWarning(ordinal: number, rule: string, quote: string | undefined): string | null {
  if (quote === undefined) return null
  const added = unquotedSpecifics(rule, quote)
  if (added.length === 0) return null
  return (
    `D${ordinal}'s rule states ${added.join(', ')}, which the operator's quote does not. ` +
    `Every digest flags it; if the operator did not say it, log the rule as they worded it with supersedes D${ordinal}.`
  )
}

File: crates/sofar-core/src/rule_fidelity.rs
//! Rule fidelity (memory-lead 1.2, D2; rust-core memory-lead 1.4) — the port
//! of `core/rule-fidelity.ts`, byte for byte against `docs/SPEC.md` §Rule
//! fidelity: what a standing rule states that the operator's own words do
//! not. A rule is the agent's restatement of what the operator said; the
//! decision may carry the operator's `quote`, and this module names the
//! SPECIFICS the rule adds — status codes and classes, paths, values.
//!
//! Pure and deterministic: no env, no clock, no locale. Text semantics are
//! JavaScript's (rust-core D2): whitespace is the `\s` class, `\w` is
//! `[A-Za-z0-9_]`, the boundary alphabet is ASCII alphanumerics, and
//! lowercasing is `String.prototype.toLowerCase` (full Unicode case mapping,
//! which `str::to_lowercase` shares for every rule this engine has seen).

use crate::text::is_js_whitespace;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SpecificKind {
    Status,
    Path,
    Value,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RuleSpecific {
    pub kind: SpecificKind,
    pub text: String,
}

/// `text.replace(/\s+/g, ' ').trim()`.
#[must_use]
pub fn collapse(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    let mut in_space = false;
    for c in text.chars() {
        if is_js_whitespace(c) {
            if !in_space {
                out.push(' ');
                in_space = true;
            }
        } else {
            out.push(c);
            in_space = false;
        }
    }
    out.trim_matches(' ').to_owned()
}

/// JS `\w`.
const fn is_word(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

const fn is_alnum(c: char) -> bool {
    c.is_ascii_alphanumeric()
}

/// `^(?:[1-5][0-9]{2}|[1-5]xx)$`, case-insensitive.
fn is_status(token: &str) -> bool {
    let b = token.as_bytes();
    if b.len() != 3 || !(b'1'..=b'5').contains(&b[0]) {
        return false;
    }
    (b[1].is_ascii_digit() && b[2].is_ascii_digit())
        || (b[1].eq_ignore_ascii_case(&b'x') && b[2].eq_ignore_ascii_case(&b'x'))
}

/// `/[\w.~-]\/[\w.*-]|^\/[\w.-]/`.
fn has_slash_path(token: &str) -> bool {
    let chars: Vec<char> = token.chars().collect();
    if chars.len() >= 2 && chars[0] == '/' && (is_word(chars[1]) || matches!(chars[1], '.' | '-')) {
        return true;
    }
    chars.windows(3).any(|w| {
        w[1] == '/'
            && (is_word(w[0]) || matches!(w[0], '.' | '~' | '-'))
            && (is_word(w[2]) || matches!(w[2], '.' | '*' | '-'))
    })
}

/// `^[\w-]{2,}(?:\.[\w-]+)*\.[A-Za-z][A-Za-z0-9]{0,4}$`.
fn is_file_name(token: &str) -> bool {
    let parts: Vec<&str> = token.split('.').collect();
    if parts.len() < 2 {
        return false;
    }
    let stem_ok = |p: &str| !p.is_empty() && p.chars().all(|c| is_word(c) || c == '-');
    let first = parts[0];
    if first.chars().count() < 2 || !stem_ok(first) {
        return false;
    }
    if !parts[1..parts.len() - 1].iter().all(|p| stem_ok(p)) {
        return false;
    }
    let ext = parts[parts.len() - 1];
    let mut it = ext.chars();
    match it.next() {
        Some(c) if c.is_ascii_alphabetic() => {}
        _ => return false,
    }
    let rest: Vec<char> = it.collect();
    rest.len() <= 4 && rest.iter().all(char::is_ascii_alphanumeric)
}

/// `^[DM][1-9][0-9]*$`.
fn is_handle(token: &str) -> bool {
    let b = token.as_bytes();
    b.len() >= 2
        && (b[0] == b'D' || b[0] == b'M')
        && (b'1'..=b'9').contains(&b[1])
        && b[2..].iter().all(u8::is_ascii_digit)
}

fn classify(token: &str) -> Option<SpecificKind> {
    if is_status(token) {
        return Some(SpecificKind::Status);
    }
    if has_slash_path(token) || is_file_name(token) {
        return Some(SpecificKind::Path);
    }
    if token.chars().any(|c| c.is_ascii_digit()) && !is_handle(token) {
        return Some(SpecificKind::Value);
    }
    None
}

const LEADING_PUNCT: &[char] = &['(', '[', '{', '<', '\'', '"', '‘', '“'];
const TRAILING_PUNCT: &[char] = &[
    ')', ']', '}', '>', '\'', '"', '’', '”', ',', ';', ':', '.', '!', '?',
];

/// The next span — backticked, straight double-quoted or curly double-quoted
/// — at or after `from`: (start, end, inner text). The regex alternation tries
/// each opener at every position from left to right, so the earliest opener
/// with a closer wins.
fn next_span(text: &str, from: usize) -> Option<(usize, usize, &str)> {
    let mut best: Option<(usize, usize, &str)> = None;
    for (open, close) in [("`", "`"), ("\"", "\""), ("“", "”")] {
        let mut search = from;
        while let Some(rel) = text[search..].find(open) {
            let start = search + rel;
            let inner_start = start + open.len();
            match text[inner_start..].find(close) {
                Some(len) if len > 0 => {
                    let end = inner_start + len + close.len();
                    if best.is_none_or(|(s, _, _)| start < s) {
                        best = Some((start, end, &text[inner_start..inner_start + len]));
                    }
                    break;
                }
                // An empty span (``) is no match at this opener; the regex moves on.
                Some(_) => search = inner_start,
                None => break,
            }
        }
    }
    best
}

/// Every specific the rule states, in rule order, deduplicated case-insensitively (`ruleSpecifics`).
#[must_use]
pub fn rule_specifics(rule: &str) -> Vec<RuleSpecific> {
    let mut found: Vec<RuleSpecific> = Vec::new();
    let mut seen: Vec<String> = Vec::new();
    let mut add = |kind: SpecificKind, text: String| {
        let key = text.to_lowercase();
        if text.is_empty() || seen.contains(&key) {
            return;
        }
        seen.push(key);
        found.push(RuleSpecific { kind, text });
    };
    let tokens_of = |chunk: &str, add: &mut dyn FnMut(SpecificKind, String)| {
        for raw in chunk.split(is_js_whitespace) {
            let token = raw
                .trim_start_matches(|c| LEADING_PUNCT.contains(&c))
                .trim_end_matches(|c| TRAILING_PUNCT.contains(&c));
            if token.is_empty() {
                continue;
            }
            if let Some(kind) = classify(token) {
                add(kind, token.to_owned());
            }
        }
    };
    let mut last = 0;
    while let Some((start, end, inner)) = next_span(rule, last) {
        tokens_of(&rule[last..start], &mut add);
        add(SpecificKind::Value, collapse(inner));
        last = end;
    }
    tokens_of(&rule[last..], &mut add);
    found
}

/// `needle` occurs in `haystack` (both already lowercased) at a term boundary.
fn contains_term(haystack: &str, needle: &str) -> bool {
    if needle.is_empty() {
        return true;
    }
    let first = needle.chars().next().is_some_and(is_alnum);
    let last = needle.chars().next_back().is_some_and(is_alnum);
    let mut from = 0;
    while let Some(rel) = haystack[from..].find(needle) {
        let at = from + rel;
        let before = first && haystack[..at].chars().next_back().is_some_and(is_alnum);
        let after = last
            && haystack[at + needle.len()..]
                .chars()
                .next()
                .is_some_and(is_alnum);
        if !before && !after {
            return true;
        }
        // `indexOf(needle, at + 1)`: one UTF-16 unit on — the next char boundary here.
        from = at + haystack[at..].chars().next().map_or(1, char::len_utf8);
        if from > haystack.len() {
            break;
        }
    }
    false
}

/// The specifics the rule states and the quote does not, as the rule spells them (`unquotedSpecifics`).
#[must_use]
pub fn unquoted_specifics(rule: &str, quote: &str) -> Vec<String> {
    let words = collapse(quote).to_lowercase();
    rule_specifics(rule)
        .into_iter()
        .filter(|s| !contains_term(&words, &collapse(&s.text).to_lowercase()))
        .map(|s| s.text)
        .collect()
}

/// `quoteClause`: `operator: "<quote>"`, plus what the rule adds to it.
#[must_use]
pub fn quote_clause(rule: &str, quote: &str) -> String {
    let added = unquoted_specifics(rule, quote);
    let flag = if added.is_empty() {
        String::new()
    } else {
        format!(" (not in the operator's words: {})", added.join(", "))
    };
    format!("operator: \"{}\"{flag}", collapse(quote))
}

/// `renderRule`: the rule verbatim (whitespace collapsed), then the operator's words when there are any.
#[must_use]
pub fn render_rule(rule: &str, quote: Option<&str>) -> String {
    let text = collapse(rule);
    match quote {
        None => text,
        Some(q) => format!("{text} — {}", quote_clause(rule, q)),
    }
}

/// `ruleFidelityWarning`: the write-time warning, or None when nothing is added.
#[must_use]
pub fn rule_fidelity_warning(ordinal: usize, rule: &str, quote: Option<&str>) -> Option<String> {
    let quote = quote?;
    let added = unquoted_specifics(rule, quote);
    if added.is_empty() {
        return None;
    }
    Some(format!(
        "D{ordinal}'s rule states {}, which the operator's quote does not. Every digest flags it; if the operator did not say it, log the rule as they worded it with supersedes D{ordinal}.",
        added.join(", ")
    ))
}

#[cfg(test)]
mod tests {
    //! The fixtures of `packages/engine/test/rule-fidelity.test.ts` (memory-lead 1.2).
    use super::*;

    const R1_RULE: &str = "Interests must be one of the provider activity categories (culture, food, nature, adventure, nightlife, wellness, tour); reject anything else with 4xx.";
    const R1_QUOTE: &str = "Reject anything else";
    const R3_RULE: &str = "Reject any interest not in the providers' Activity category set (apps/web/lib/categories.ts).";

    fn specifics(rule: &str) -> Vec<(SpecificKind, &str)> {
        rule_specifics(rule)
            .into_iter()
            .map(|s| (s.kind, Box::leak(s.text.into_boxed_str()) as &str))
            .collect()
    }

    #[test]
    fn finds_the_round_1_addition_and_nothing_the_operator_said() {
        assert_eq!(unquoted_specifics(R1_RULE, R1_QUOTE), vec!["4xx"]);
    }

    #[test]
    fn classifies_status_codes_paths_and_values() {
        assert_eq!(
            specifics(
                "Return 404, never 5xx, from /api/chat in apps/web/route.ts within 30s, max `retries: 3`"
            ),
            vec![
                (SpecificKind::Status, "404"),
                (SpecificKind::Status, "5xx"),
                (SpecificKind::Path, "/api/chat"),
                (SpecificKind::Path, "apps/web/route.ts"),
                (SpecificKind::Value, "30s"),
                (SpecificKind::Value, "retries: 3"),
            ]
        );
    }

    #[test]
    fn ignores_prose_abbreviations_and_record_handles() {
        assert!(
            rule_specifics("Keep SQLite, e.g. for tests, as D3 and M2 say; i.e. nothing else.")
                .is_empty()
        );
    }

    #[test]
    fn a_double_quoted_span_is_one_value_never_re_read_as_tokens() {
        assert_eq!(
            specifics("Label the button \"Save 2 drafts\" only"),
            vec![(SpecificKind::Value, "Save 2 drafts")]
        );
        assert_eq!(
            specifics("Label it “Save 2 drafts” only"),
            vec![(SpecificKind::Value, "Save 2 drafts")]
        );
    }

    #[test]
    fn matches_case_insensitively_at_term_boundaries() {
        assert!(
            unquoted_specifics(
                "Return 400 for API/v2 errors",
                "return 400 for api/v2 errors"
            )
            .is_empty()
        );
        assert_eq!(unquoted_specifics("Return 400", "port 4000"), vec!["400"]);
    }

    #[test]
    fn a_path_the_operator_never_said_is_flagged() {
        assert_eq!(
            unquoted_specifics(R3_RULE, R1_QUOTE),
            vec!["apps/web/lib/categories.ts"]
        );
    }

    #[test]
    fn render_and_warning_follow_the_engine() {
        assert_eq!(
            render_rule(R1_RULE, Some(R1_QUOTE)),
            format!("{R1_RULE} — operator: \"{R1_QUOTE}\" (not in the operator's words: 4xx)")
        );
        assert_eq!(
            render_rule("Reject anything else", Some(R1_QUOTE)),
            "Reject anything else — operator: \"Reject anything else\""
        );
        assert_eq!(render_rule("  two   spaces ", None), "two spaces");
        assert_eq!(
            rule_fidelity_warning(2, R1_RULE, Some(R1_QUOTE)).as_deref(),
            Some(
                "D2's rule states 4xx, which the operator's quote does not. Every digest flags it; if the operator did not say it, log the rule as they worded it with supersedes D2."
            )
        );
        assert_eq!(rule_fidelity_warning(2, R1_RULE, None), None);
        assert_eq!(
            rule_fidelity_warning(2, "Reject anything else", Some(R1_QUOTE)),
            None
        );
    }

    #[test]
    fn token_edges_and_spans() {
        assert_eq!(
            specifics("(see /api/chat)."),
            vec![(SpecificKind::Path, "/api/chat")]
        );
        assert_eq!(
            specifics("v2 and D12 and M3 and 3rd"),
            vec![(SpecificKind::Value, "v2"), (SpecificKind::Value, "3rd")]
        );
        assert_eq!(
            specifics("Use `x` then `X` then 4XX"),
            vec![(SpecificKind::Value, "x"), (SpecificKind::Status, "4XX")]
        );
        assert!(rule_specifics("a.b").is_empty());
        assert_eq!(
            specifics("ab.ts and file.tar.gz and x.toolong"),
            vec![
                (SpecificKind::Path, "ab.ts"),
                (SpecificKind::Path, "file.tar.gz")
            ]
        );
    }
}
